crates/prune/src/segments/headers.rs

```diff
@@ -36,90 +36,174 @@ impl<DB: Database> Segment<DB> for Headers {
-        let mut last_pruned_block = *range.end();
-        let (pruned, done) = provider.prune_table_with_range::<T>(
-            range,
-            delete_limit,
-            |_| false,
-            |row| last_pruned_block = row.0,
-        )?;
-        trace!(target: "pruner", %pruned, %done, table = %T::TABLE, "Pruned headers");
-
-        Ok((done, pruned, last_pruned_block))
+#[allow(missing_debug_implementations)]
+struct HeaderTablesIter<'a, 'b, DB>
+where
+    DB: Database,
+{
+    provider: &'a DatabaseProviderRW<DB>,
+    limiter: &'b mut PruneLimiter,
+    last_pruned_block: Option<u64>,
+    to_block: u64,
+}
+
+impl<'a, 'b, DB> HeaderTablesIter<'a, 'b, DB>
+where
+    DB: Database,
+{
+    fn new(
+        provider: &'a DatabaseProviderRW<DB>,
+        limiter: &'b mut PruneLimiter,
+        last_pruned_block: Option<u64>,
+        to_block: u64,
+    ) -> Self {
+        Self { provider, limiter, last_pruned_block, to_block }
+    }
+}
+
+impl<'a, 'b, DB> Iterator for HeaderTablesIter<'a, 'b, DB>
+where
+    DB: Database,
+{
+    type Item = Result<Option<u64>, PrunerError>;
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self { provider, limiter, last_pruned_block, to_block } = self;
+
+        if limiter.is_limit_reached() || Some(*to_block) == *last_pruned_block {
+            return None
+        }
+
+        let block_step =
+            if let Some(block) = *last_pruned_block { block + 1..=block + 2 } else { 0..=1 };
+
+        let next_up_last_pruned_block = Some(*block_step.start());
+        let mut last_pruned_block_headers = None;
+        let mut last_pruned_block_td = None;
+        let mut last_pruned_block_canonical = None;
+        // todo: guarantee skip filter and delete callback are same for all header table types
+
+        if let Err(err) =
+            provider.with_walker::<tables::Headers, _, _>(block_step.clone(), |ref mut walker| {
+                provider.step_prune_range(walker, limiter, &mut |_| false, &mut |row| {
+                    last_pruned_block_headers = Some(row.0)
+                })
+            })
+        {
+            return Some(Err(err.into()))
+        }
+
+        if let Err(err) = provider.with_walker::<tables::HeaderTerminalDifficulties, _, _>(
+            block_step.clone(),
+            |ref mut walker| {
+                provider.step_prune_range(walker, limiter, &mut |_| false, &mut |row| {
+                    last_pruned_block_td = Some(row.0)
+                })
+            },
+        ) {
+            return Some(Err(err.into()))
+        }
+
+        if let Err(err) =
+            provider.with_walker::<tables::CanonicalHeaders, _, _>(block_step, |ref mut walker| {
+                provider.step_prune_range(walker, limiter, &mut |_| false, &mut |row| {
+                    last_pruned_block_canonical = Some(row.0)
+                })
+            })
+        {
+            return Some(Err(err.into()))
+        }
```

Comment:
as far as I understand, it means that we open a new cursor on every iteration. It's not good and adds a significant overhead, we should rather initialize the cursors once on `::new(...)` and then advance them on each iteration.